test/lit/passes/ldse.wast

```diff
@@ -0,0 +1,1481 @@
+ ;; CHECK-NEXT: )
+ (func $no-basic-blocks
+  (unreachable)
+ )
+
+ ;; CHECK:      (func $global
+ ;; CHECK-NEXT:  (drop
+ ;; CHECK-NEXT:   (i32.const 10)
+ ;; CHECK-NEXT:  )
+ ;; CHECK-NEXT:  (global.set $global$1
+ ;; CHECK-NEXT:   (i32.const 20)
+ ;; CHECK-NEXT:  )
+ ;; CHECK-NEXT:  (global.set $global$0
+ ;; CHECK-NEXT:   (i32.const 30)
+ ;; CHECK-NEXT:  )
+ ;; CHECK-NEXT: )
+ (func $global
+  ;; globals are optimized as well, and we have more precise data there than on
+  ;; GC references - aliasing is impossible, and so we can tell this first one
+  ;; is dead due to the last, ignoring the unaliasing one in the middle
+  (global.set $global$0
+   (i32.const 10)
+  )
+  (global.set $global$1
+   (i32.const 20)
+  )
+  (global.set $global$0
+   (i32.const 30)
+  )
+ )
+
+ ;; CHECK:      (func $global-trap
+ ;; CHECK-NEXT:  (global.set $global$0
+ ;; CHECK-NEXT:   (i32.const 10)
+ ;; CHECK-NEXT:  )
+ ;; CHECK-NEXT:  (if
+ ;; CHECK-NEXT:   (i32.const 1)
+ ;; CHECK-NEXT:   (unreachable)
+ ;; CHECK-NEXT:  )
+ ;; CHECK-NEXT:  (global.set $global$0
+ ;; CHECK-NEXT:   (i32.const 20)
+ ;; CHECK-NEXT:  )
+ ;; CHECK-NEXT: )
+ (func $global-trap
+  (global.set $global$0
+   (i32.const 10)
+  )
+  ;; a trap (even conditional) prevents our optimizations, global state may be
+  ;; observed if another export is called later after the trap.
+  (if
+   (i32.const 1)
+   (unreachable)
+  )
+  (global.set $global$0
+   (i32.const 20)
+  )
+ )
+
+ ;; CHECK:      (func $memory-const
+ ;; CHECK-NEXT:  (block
+ ;; CHECK-NEXT:   (drop
+ ;; CHECK-NEXT:    (i32.const 10)
+ ;; CHECK-NEXT:   )
+ ;; CHECK-NEXT:   (drop
+ ;; CHECK-NEXT:    (i32.const 20)
+ ;; CHECK-NEXT:   )
+ ;; CHECK-NEXT:  )
+ ;; CHECK-NEXT:  (i32.store
+ ;; CHECK-NEXT:   (i32.const 10)
+ ;; CHECK-NEXT:   (i32.const 30)
+ ;; CHECK-NEXT:  )
+ ;; CHECK-NEXT: )
+ (func $memory-const
+  ;; test dead store elimination of writes to memory at constant offsets
+  (i32.store
+   (i32.const 10)
+   (i32.const 20)
+  )
+  (i32.store
+   (i32.const 10)
+   (i32.const 30)
+  )
+ )
+
+ ;; CHECK:      (func $memory-param (param $x i32)
+ ;; CHECK-NEXT:  (block
+ ;; CHECK-NEXT:   (drop
+ ;; CHECK-NEXT:    (local.get $x)
+ ;; CHECK-NEXT:   )
+ ;; CHECK-NEXT:   (drop
+ ;; CHECK-NEXT:    (i32.const 20)
+ ;; CHECK-NEXT:   )
+ ;; CHECK-NEXT:  )
+ ;; CHECK-NEXT:  (i32.store
+ ;; CHECK-NEXT:   (local.get $x)
+ ;; CHECK-NEXT:   (i32.const 30)
+ ;; CHECK-NEXT:  )
+ ;; CHECK-NEXT: )
+ (func $memory-param (param $x i32)
+  ;; test dead store elimination of writes to memory using a local
+  (i32.store
+   (local.get $x)
+   (i32.const 20)
+  )
+  (i32.store
+   (local.get $x)
+   (i32.const 30)
+  )
+ )
+
+ ;; CHECK:      (func $memory-wrong-const
+ ;; CHECK-NEXT:  (i32.store
+ ;; CHECK-NEXT:   (i32.const 10)
+ ;; CHECK-NEXT:   (i32.const 20)
+ ;; CHECK-NEXT:  )
+ ;; CHECK-NEXT:  (i32.store
+ ;; CHECK-NEXT:   (i32.const 30)
+ ;; CHECK-NEXT:   (i32.const 40)
+ ;; CHECK-NEXT:  )
+ ;; CHECK-NEXT: )
+ (func $memory-wrong-const
```

Comment:
Nit: How about 'different` instead of 'wrong'? The same for other function names.